kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaEvictionService.java

```diff
@@ -0,0 +1,164 @@
+package com.slack.kaldb.clusterManager;
+
+import static com.google.common.util.concurrent.Futures.addCallback;
+import static com.slack.kaldb.config.KaldbConfig.DEFAULT_ZK_TIMEOUT_SECS;
+import static com.slack.kaldb.util.FutureUtils.successCountingCallback;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.AbstractScheduledService;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.slack.kaldb.metadata.cache.CacheSlotMetadata;
+import com.slack.kaldb.metadata.cache.CacheSlotMetadataStore;
+import com.slack.kaldb.metadata.replica.ReplicaMetadata;
+import com.slack.kaldb.metadata.replica.ReplicaMetadataStore;
+import com.slack.kaldb.proto.config.KaldbConfigs;
+import com.slack.kaldb.proto.metadata.Metadata;
+import io.micrometer.core.instrument.Counter;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Timer;
+import java.time.Instant;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Marks expired replicas that are currently assigned to a cache slot for eviction. This service
+ * does not delete or mutate replicas, but only marks the cache slot as pending eviction. Once the
+ * cache node has evicted an expired replica, a separate service will delete expired, unassigned
```

Comment:
Why do we need a separate service for replica deletion? Since, the logic to evict replicas is here, is this the best place to know what replicas to delete?

Reply:
As discussed, we're keeping these as separate services for two reasons - one is that eviction can take a while to complete, and the second is we want replica eviction and replica deletion to be independently fault tolerant. We could combine these into a single service / class, but as they're doing independent operations it makes sense to keep them separate for now.